Open.h

```diff
@@ -3,16 +3,206 @@
 #include "HashTable.h"
 #include "Product.h"
 #include <iostream>
-#include <map>
+#include <vector>
+#include <cmath>
+#include <string>
 
 template<typename Key, typename Value>
 class OpenMap: public HashTable<Key, Value> {
-    public:
-        OpenMap() = default;
+private:
+    double MAX_LOAD_FACTOR = 0.7;
 
-        Value get(Key k) = 0;
+    struct KeyValuePair{
+        KeyValuePair(Key k, Value v) {
+            key = k;
+            value = v;
+            valid = true;
+        }
+        KeyValuePair(){
+            key = Key{};
+            value = Value{};
+            valid = false;
+        }
+        Key key;
+        Value value;
+        bool valid;
+    };
 
-        void put(Key k, Value v) = 0;
+    //this is the operation done on any index that collides
+    uint64_t iterate(uint64_t index){
+        return index*index;
+    }
 
-        void remove(Key k) = 0;
+    std::vector<KeyValuePair>* hashmap;
+    std::vector<bool>* significancemap;
+
+    std::hash<Key> hasher;
+
+    uint64_t hash(Key k){
+        return hasher(k)%hashmapSize;
+    }
+
+    uint64_t hashmapSize = 4;
+
+    uint64_t numPairs = 0;
+    uint64_t numSignificant = 0;
+
+    void doubleSize(){
+        hashmapSize*= 2;
+        auto newMap = new std::vector<KeyValuePair>(hashmapSize);
+        auto newSignificance = new std::vector<bool>(hashmapSize);
+        for(auto kvp : (*hashmap)){
+            if(kvp.valid){
+                uint64_t index = hash(kvp.key);
+                while((*newSignificance)[index%hashmapSize]){
+                    index = iterate(index);
+                }
+                (*newMap)[index%hashmapSize] = kvp;
+                (*newSignificance)[index%hashmapSize] = true;
+            }
+        }
+        numSignificant = numPairs;
+        //clear old memory
+        auto temp = hashmap;
+        auto temp2 = significancemap;
+        hashmap = newMap;
+        significancemap = newSignificance;
+        delete temp;
+        delete temp2;
+    }
+
+    Value defaultGetReturn(){
+        std::cerr << "Invalid types for defaultGetReturn";
+        exit(1);
+    }
+
+    void printKeyValuePair(KeyValuePair kvp){
+        std::cout << " Invalid types for printing ";
+    }
+
+public:
+    OpenMap() {
+        hashmap =  new std::vector<KeyValuePair>(hashmapSize);
+        significancemap = new std::vector<bool>(hashmapSize, false);
+    }
+
+    OpenMap(double load) {
+        if(load <= 0 || load >= 1){
+            std::cerr << "Invalid load factor for OpenGraph (not 0..1) - setting to 0.7";
+            MAX_LOAD_FACTOR = 0.7;
+        }
+        else{
+            MAX_LOAD_FACTOR = load;
+        }
+        hashmap =  new std::vector<KeyValuePair>(hashmapSize);
+        significancemap = new std::vector<bool>(hashmapSize, false);
+    }
+
+    ~OpenMap(){
+        delete hashmap;
+        delete significancemap;
```

Comment:
It is good that you are deleting your new'ed memory, but best to never call new to begin with.

Reply:
Also, destructor order in c++ is not defined, so keep them as simple as possible.  They can really screw you up if they do anything non-trivial like call into a library.